### DISCUSSION TITLE
Do I understand processing dynamic scope correctly?

### DISCUSSION THREAD

**Discussion by flaksp** [json-schema-org/community#264] [Q&A] [answered]
Hello!

I'm trying to understand dynamic scope and references better. I want to be sure I understand it correctly using following [official test case](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/104f9f98f782d258eab6b2a760f9dca41739cebe/tests/draft2020-12/dynamicRef.json#L120-L160) as an example:

```json
{
  "description": "A $dynamicRef with intermediate scopes that don't include a matching $dynamicAnchor does not affect dynamic scope resolution",
  "schema": {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": "https://test.json-schema.org/root",
    "$ref": "intermediate-scope",
    "$defs": {
      "foo": {
        "$dynamicAnchor": "items",
        "type": "string"
      },
      "intermediate-scope": {
        "$id": "intermediate-scope",
        "$ref": "list"
      },
      "list": {
        "$id": "list",
        "type": "array",
        "items": {
          "$dynamicRef": "#items"
        },
        "$defs": {
          "items": {
            "$comment": "This is only needed to satisfy the bookending requirement",
            "$dynamicAnchor": "items"
          }
        }
      }
    }
  },
  "tests": [
    {
      "description": "An array of strings is valid",
      "data": [
        "foo",
        "bar"
      ],
      "valid": true
    },
    {
      "description": "An array containing non-strings is invalid",
      "data": [
        "foo",
        42
      ],
      "valid": false
    }
  ]
}
```

My implementation starts processing the document step-by-step, which produces following **stack** (inspired by "[JSON Schema as a Framework](https://github.com/orgs/json-schema-org/discussions/220)" slides, part 2):

1. Processing schema root:
    - Schema location: `https://test.json-schema.org/root#`
    - Evaluation path: (empty)
2. It meets `$ref` keyword:
    - Schema location: `https://test.json-schema.org/intermediate-scope#`
    - Evaluation path: `/$ref`
3. It meets `$ref` keyword again:
    - Schema location: `https://test.json-schema.org/list#`
    - Evaluation path: `/$ref/$ref`
4. It goes into `items`:
    - Schema location: `https://test.json-schema.org/list#/items`
    - Evaluation path: `/$ref/$ref/items`

After that, it meets `$dynamicRef` keyword with `#items` value.

As I understand from the slides, my implementation should go back down **the stack**:

1. Schema location is `https://test.json-schema.org/list#/items`. There is a schema `https://test.json-schema.org/list#items` defined by `$dynamicAnchor` ✅
2. Schema location is `https://test.json-schema.org/list#`. There is a schema `https://test.json-schema.org/list#items` defined by `$dynamicAnchor` ✅
3. Schema location is `https://test.json-schema.org/intermediate-scope#`. There is NO schema `https://test.json-schema.org/intermediate-scope#items` ⛔️
4. Schema location is `https://test.json-schema.org/root#`. There is a schema `https://test.json-schema.org/root#items` defined by `$dynamicAnchor` ✅

Stack is over. The last dynamic schema in the stack it found is `https://test.json-schema.org/root#items`, that's why it should be picked. It is correct?

**Comment by handrews** [comment, marked as answer]
Yes, you want the outermost (first stack frame/dynamic scope encountered) match, and that is the `.../root#items`

**Reply by flaksp** [reply]
How will it behave if `$dynamicRef` contains not just plain fragment, but something different, e.g. URI reference or JSON Pointer (or both)? I don't see such test cases 🤔

**Reply by handrews** [reply]
`$dynamicRef` to something that does not include a plain name fragment works just like `$ref`, as does one to a URI that includes a plain name fragment defined by `$anchor` instead of by `$dynamicAnchor`.

If you have something like `"$dynamicRef": "/schemas/foo#bar"` and `bar` in `/schemas/foo` was defined by `$dynamicAnchor`, all this affects is the starting point.  So when you start unwinding the stack, you start at `<whatever base URI>/schemas/foo#bar`.  You would do this if your current schema resource doesn't have the right `$dynamicAnchor` and you want to target another schema resource as the fallback resolution (meaning that if there's nothing else matching in the stack, it falls back to that starting point).
